server/decoders/character.ts

```diff
@@ -0,0 +1,38 @@
+// ---------------------------------------------------------------------------------------------------------------------
+// Character Decoders
+// ---------------------------------------------------------------------------------------------------------------------
+
+import { inexact, object, optional, string } from 'decoders';
+import { hexColor, jsonObjectString, nullToUndefined, stringWithLength, withDefault } from './utils';
+
+// ---------------------------------------------------------------------------------------------------------------------
+
+export const characterRecDecoder = object({
+    id: string,
+    system: stringWithLength(3, 255),
+    accountID: stringWithLength(3, 255),
+    noteID: stringWithLength(3, 255),
+    name: stringWithLength(1, 255),
+    description: optional(string),
+    portrait: optional(nullToUndefined(string)),
+    thumbnail: optional(string),
+    color: optional(hexColor),
+    campaign: optional(stringWithLength(0, 255)),
+    details: jsonObjectString(inexact({}))
```

Comment:
A slightly more flexible way to do this is to pass in the detail decoder. This means when creating the characters from rec/json for use with systems that want the details, the decoder can verify the stuff itself. If I remember right typescript allows default values in the parameter list, so you can still keep it option.

```typescript
const defaultDetailsDecoder = jsonObjectString(inexact({}));

export const charactRecDecoder = function(detailDecoder){
    if(isUndefined(detailDecoder)){
        detailDecoder = defaultDetailDecoder;
    }
    return object({...details: detailDecoder});
}
```

Reply:
Ok, that's not bad. I didn't think about dynamically building the decoder, that's clever. But, since the type is a property of the outer object, don't I first need to decode the type, and then I can use it to build the details?